### PR TITLE
Proper Refactor

### PR DESCRIPTION
**Pull request by ItsANoBrainer**
Refactored the entire resource because the key system was not working.

No longer does a server callback 10 or so times a second, on top of other bad practices.

Tested pretty much everything with my devs, but there might be some weird issues we havnt found. Please lookover as I've been going crazy with the small tedious issues I've come across doing this.

Throwing this up to get feedback, and for people to use who want it.

Features:

- Keys are saved server side, and sent to each client when they are added or removed, as well as on character selection. This allows characters to leave and come back in the same server uptime session and still have the same keys
- Keys are only pulled from the server once on character load
- Giving keys has 3 types, /givekeys with an id will attempt to give keys you have to an id, not including an id will try to give it to the closest person, and if you're in a vehicle it will give to everyone
- Can only give keys to vehicles you have keys to
- Server event to force acquire keys to a plate for a person (for lockpicking/stealing/spawning, etc.)
- Can Carjack an NPC with a gun, has different percentages to work based on weapon (config)
- Take keys from dead npc drivers
- Can lockpick a car to unlock the doors
- Can hotwire a car if you're in the driver seat of the vehicle you dont have keys to to get keys
- Keeps the engine of a vehicle off if you dont have keys to it (allows other resources to attempt to toggle car engine without needing to interface if you have keys or not)
- Vehicle blacklist system for vehicles you dont want to be lockable (always unlocked)
- Police Alerts from lockpicking/hotwiring/carjacking
- Export to check if you have keys to a vehicle
- Locking/unlocking with hotkey
- Locking/unlocking and giving keys uses a custom GetVehicleInDirection you are facing for realism/accuracy
- Fully compatible with old resource. Has an event handler for 'vehiclekeys:client:SetOwner' in which it gives keys for that plate
- Some other shit I probably forgot

TODO:
- Add LOCALE support
- Stop peds from wanting to get back into the vehicle they just got carjacked out of as they're fleeing

### CODE REVIEW

**Comment by ItsANoBrainer** [comment]
Works perfectly fine for me, all doors are locked.

**Comment by Andyauk** [comment]
> Works perfectly fine for me, all doors are locked
> 
>  FiveM_b2545_GTAProcess_8VDvfnfadA.mp4 
> .

ill redownload and try again. thanks

File: client/main.lua [hunk 1]
```diff
@@ -1,287 +1,563 @@
...
+            sprite = data.sprite,
+            scale = 1.2,
+            colour = 3,
+            flashes = false,
+            title = '10-60 - Vehicle Theft',
+            time = (5*60*1000),
+            sound = 1,
+        }
+    })
+end
```
Comment (menkzy) [inline client/main.lua:537]: Calling cd_dispatch, should be calling the default built in QBCore alerts (although not many really use it lul)

**Comment by Andyauk** [comment]
don't get me wrong, cd_dispatch is great, but 60 bucks, is too much on it.

File: client/main.lua [hunk 2]
```diff
@@ -1,287 +1,563 @@
...
         TriggerServerEvent('hud:server:GainStress', math.random(1, 4))
-        QBCore.Functions.Notify('Someone Called The Police!', 'error')
+        AttemptPoliceAlert("steal")
     end
+
     if usingAdvanced then
         if chance <= Config.RemoveLockpickAdvanced then
-            TriggerEvent('inventory:client:ItemBox', QBCore.Shared.Items["advancedlockpick"], "remove")
-            TriggerServerEvent("QBCore:Server:RemoveItem", "advancedlockpick", 1)
+            TriggerServerEvent("inventory:server:breakLockpick", "advancedlockpick")
```
Comment (TonybynMp4) [inline client/main.lua:386]: There's no breakLockpick event in qb-inventory

**Comment by davidxd33** [comment]
Is there an update to this PR?

**Comment by MatthewRorke** [comment]
I brought this in and made minor adjustments to the dispatcher to use the one we use on our city, and I am now using this codebase successfully in production.

Works for me, 10/10

**Comment by ADRNALN** [comment]
Any reason the distance check was removed for keys? Players have to be right up against a vehicle now to unlock/lock

**Comment by MatthewRorke** [comment]
> Any reason the distance check was removed for keys? Players have to be right up against a vehicle now to unlock/lock

No they don't, they need to face the vehicle.

**Pull request merged**